### PR TITLE
Add F# support

### PR DESCRIPTION
**Pull request by srivatsn**
This PR sets up the skeleton for what's needed to add basic F# support.

This change adds a new project system with a new project type guid. A IVsProjectSelector needs to be implemented to select this guid for a given project based on some condition of the project. To test though, the .sln file can be modified with this new guid and the project will open with the new project system.

Designtime targets: There's a new file called Microsoft.FSharp.DesignTime.targets that needs to be pulled in by the FSharp targets similar to how it's done for C# [here](https://github.com/Microsoft/msbuild/blob/xplat/src/Tasks/Microsoft.CSharp.CurrentVersion.targets#L325). This is so that the right CPS rules and capabilities get added to the project.

F# capability: The Designtime.targets adds a new FSharp capability. However all of the project system components today are marked as applicable for CSharp or VisualBasic. We need to go through them all and mark them as applicable to F# as well where it makes sense.

Language-service-hookup. There's three pieces of work here:
- The way the LS hookup works in roslyn project system is that kicks off a designtime build of a target called [CompileDesignTime](https://github.com/dotnet/roslyn-project-system/compare/master...srivatsn:fsharp?expand=1#diff-87271b4777d84b19f2d66ace1b99ac28R39) and expects to get the commandline string that would have been passed to the compiler. To do that the Fsc task needs to return the FscCommandLineArgs similar to how the csc task does that [here](https://github.com/dotnet/roslyn/blob/e509b57971474c4d5937bb9a7434839ee8e2c31b/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L142)
- Once the commandline is obtained, it needs to be parsed and FSharpCommandLineParserProvider needs to be filled out by calling into the F# compiler.
- Finally, we need to create a ProjectContext. There's common code that does that today. However it looks like Roslyn has a check for C#\VB in that code path and a different overload was added for F#. We should just fix that and standardize on calling with an error prefix.

Then there's the big workitem to show the items ordered by fileorder in the tree and provide gestures to move them around. 

I believe @brettfo @KevinRansom are going to fill out some of the pieces here.

@dotnet/project-system @Pilchie

### CODE REVIEW

**Comment by srivatsn** [comment]
fs

**Comment by forki** [comment]
Kiss

**Comment by forki** [comment]
I'm really excited for this. Thanks for doing this. 

One question after seeing the screenshot: will the package references work like in C#? I mean will it be possible to fill these in memory? This is important in order to keep paket dependency manager support. @davkean I assume yes, but would be nice to hear that everything will be the same.
Thx

**Comment by davkean** [comment]
Package references will work exactly like C#.

**Comment by forki** [comment]
Awesome thanks for confirmation.

**Comment by enricosada** [comment]
> To do that the Fsc task needs to return the FscCommandLineArgs similar to how the csc task does that here

Already done that with https://github.com/dotnet/netcorecli-fsc/pull/88 (FSharp.NET.Sdk, the target file) and https://github.com/Microsoft/visualfsharp/pull/2523 (Fsc task)

@srivatsn if you want, you can use test it with the fsproj and nuget.config in https://github.com/enricosada/dotnet-proj-info/tree/master/examples as example. 
The nuget.config contains the dev feed with new fsctask and fsharp.net.sdk
the fsproj is the new tempalte for 1.0.0 rtm, so without dotnet-compile-fsc

i use that for a dotnetcli tool who will get fsc args `dotnet proj-info -p ..\c1\c1.fsproj --fsc-args`, ref https://github.com/enricosada/dotnet-proj-info/tree/master/examples

**Comment by srivatsn** [comment]
@enricosada is there a plan to add that back to the regular desktop Fsc task as well? This project system should eventually work for all F# projects.

**Comment by enricosada** [comment]
@srivatsn the package Fsharp.Net.Sdk version 1.0.0 already use only Fsc task (not dotnet compile fsc). Is the one from vf# repo. 

And also the new fsproj already works to create lib/console target framework net451 etc 

Works on mono too :)

**Comment by enricosada** [comment]
And net45 are under integration test suite in https://github.com/dotnet/netcorecli-fsc/blob/master/test/dotnet-new.Tests/CommonScenarioTests.cs#L55


@srivatsn or you meant something different?

**Comment by srivatsn** [comment]
Sorry I should have been clearer - the fact that the FscTask supports this now is great. However, I was referring to https://github.com/dotnet/netcorecli-fsc/pull/88. My understanding is that that PR is updating the FSharp.targets that are used in the FSharp.NET.Sdk which means any project using the SDK attribute and pulling that SDK will get this. However, old projects that exist today (created using previous versions of VS targeting desktop) will not create the commandline args. I think we should replace the existing project system with this effort and get all F# projects to open up with this new project system which would mean fixing up the targets in the [VF# repo](https://github.com/Microsoft/visualfsharp/blob/2d258b1209e88fedf29b46e6f0da6297de1ca4be/src/fsharp/FSharp.Build/Microsoft.FSharp.targets) - correct? So I was asking if you plan to PR those changes into the VF# repo.

To put this differently, I see that the F# SDK enables a bunch more things and i'm wondering if we should consolidate that with the targets in the VF# repo so that we don't have two styles of projects. For eg, one of the other workitems I mention above - adding an import the FSharp.DesignTime.targets that I'm adding in this PR needs to be done in the F# SDK and in the F# targets in the VF# repo before importing common.targets.

**Comment by enricosada** [comment]
An update:

`FSharp.Compiler.Tools` 4.1.1 (used by `FSharp.NET.Sdk` 1.0.2) add the new properties to embedded `Fsc` task.
Eveything is on nuget.org, and the `dotnet new` templates (of .net core tools 1.0.1) resolve that now.

Ready to use examples also in https://github.com/enricosada/DotnetNewFsprojTestingSamples to test p2p and packageref

**Comment by srivatsn** [comment]
Thanks @enricosada - I just tried the FSharp.NET.Sdk 1.0.2. There's still a problem. It looks like the new properties have been added to the netcoreapp1.0 version of the Fsc task in FSharp.Build.dll but not to the desktop version. Since VS uses desktop msbuild, it picks up the Fsharp.Build.dll in the tools folder in that package and fails design time builds fail saying it can't find the SkipCompilerExecution property.

**Comment by srivatsn** [comment]
@dotnet/project-system I think this is ready for review. I've cleaned up some of the hacks, added tests and created issues for the remaining work - 

#1853 - F# language service integration 
#1854 - F# property pages 
#1855 - F# item templates don't show up 
#1856 - F# resx and xaml files need to be localized and setup authored. 
#1857 - F# SDK needs to import design-time targets and authored into VS

File: src/Targets/Microsoft.FSharp.DesignTime.targets [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    <PropertyPageSchema Include="$(ManagedXamlResourcesDirectory)FSharp.BrowseObject.xaml;">
+      <Context>BrowseObject</Context>
+    </PropertyPageSchema>
+
+    <ProjectCapability Include="FSharp;Managed"/>
+  </ItemGroup>
+
+  <!-- Targets -->
+
+  <!-- Returns Csc command-line arguments for the language service -->
```
Comment (saul) [inline src/Targets/Microsoft.FSharp.DesignTime.targets]: Csc -> Fsc

**Comment by saul** [comment]
@srivatsn this looks great, thanks for doing this :)

Out of interest, what needs to be completed from the F# front? i.e. is there any support for arbitrary ordering in the solution explorer? I'm sure this has been mentioned before but I can't remember where I saw it.

**Comment by enricosada** [comment]
> I just tried the FSharp.NET.Sdk 1.0.2. There's still a problem. It looks like the new properties have been added to the netcoreapp1.0 version of the Fsc task in FSharp.Build.dll but not to the desktop version. Since VS uses desktop msbuild, it picks up the Fsharp.Build.dll in the tools folder in that package and fails design time builds fail saying it can't find the SkipCompilerExecution property

yes, is an issue also for mono. Is tracked in https://github.com/fsharp/fsharp/issues/684

**Comment by srivatsn** [comment]
> is there any support for arbitrary ordering in the solution explorer?
No, that needs to be implemented as well. Filed https://github.com/dotnet/roslyn-project-system/issues/1875 to track that.

**Comment by saul** [comment]
Thanks @srivatsn. 

What's the easy way to get setup and working on these tickets? I'm happy to do some of the work, especially as I just spent a few days working on the file ordering in the old project system.

What's your dev cycle at the minute?

cc @brettfo - are these tickets open to be worked on by the community at this point, or just MS?

**Comment by srivatsn** [comment]
@saul - sure feel free to pick up any of this work. Let @brettfo know since I believe he plans to tackle them otherwise. https://github.com/dotnet/roslyn-project-system/blob/master/docs/repo/getting-started.md has instructions to get started.

**Comment by brettfo** [comment]
Also to @saul (and @srivatsn), one of the steps will be to provide an F# command line parser.  I've created #1876 which starts this work, but that's ultimately dependent upon some changes in Roslyn.  Once both of these are merged, I'll start work on an F# command line parser that can be used here.

**Comment by cloudRoutine** [comment]
@brettfo like this one?
https://github.com/cloudRoutine/netcorecli-fsc/blob/fsharp-impl/src/dotnet-compile-fsc/Program.fs
https://github.com/cloudRoutine/netcorecli-fsc/tree/fsharp-impl/src/dotnet-compile-fsc

or does it need to use a different API?

**Comment by brettfo** [comment]
@saul, Speaking with @KevinRansom he suggested factoring out [this](https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/fsc.fs#L1635) method.  My plan was to add a type `FSharpCommandLineArguments that inherts from the Roslyn type `CommandLineArguments`, but first some APIs need to be opened up (see here: dotnet/roslyn#18258).

Once that's been done I need to consume the F# command line parser [here](https://github.com/dotnet/roslyn-project-system/pull/1876/commits/d82b0087f7bd263e9eb6ba7eff5f1a04c03cd77c).

**Comment by saul** [comment]
@brettfo @srivatsn do you have an example project file?

I've followed the instructions (using RoslynDev suffix) and I'm using one of @enricosada's example F# .Net Core projects, but I'm getting the following error:

image

**Comment by srivatsn** [comment]
@saul Until https://github.com/dotnet/roslyn-project-system/issues/1857 is fixed, you'll need to make those changes in your machine. That is, you need to:
- Copy the FSharp.NET.Sdk folder from %Program Files%\Dotnet\SDK\sdks to %vsdir%\msbuild\Sdks
- Add the snippet of xml mentioned in #1857 to the two FSharp.NET.Sdk (one in the VSDIR and the other one in your nuget packages directory).
- Then you can open a F# netcore project by either renaming to msbuildproj as you seem to be doing or by changing the project type guid in the .sln file from `F2A71F9B-5D33-465A-A702-920D77279786` (Existing project system's guid) to `6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705` (new project system's guid)

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Packaging/FSharpProjectSystemPackage.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+using Microsoft.VisualStudio.ProjectSystem;
+
+// We register ourselves as a new CPS "project type"
+[assembly: ProjectTypeRegistration(projectTypeGuid: FSharpProjectSystemPackage.ProjectTypeGuid, displayName: "#1", displayProjectFileExtensions: "#2", defaultProjectExtension: "fsproj", language: "FSharp", resourcePackageGuid: FSharpProjectSystemPackage.PackageGuid, Capabilities = ManagedProjectSystemPackage.DefaultCapabilities + "; " + ProjectCapability.FSharp, DisableAsynchronousProjectTreeLoad = true)]
+
+namespace Microsoft.VisualStudio.Packaging
+{
+    [Guid(PackageGuid)]
+    [PackageRegistration(AllowsBackgroundLoading = true, RegisterUsing = RegistrationMethod.CodeBase, UseManagedResourcesOnly = true)]
+    [RemoteCodeGeneratorRegistration(SingleFileGenerators.ResXGuid, SingleFileGenerators.ResXGeneratorName,
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Packaging/FSharpProjectSystemPackage.cs]: It looks like you got these off C#, are these the same ones that are enabled for existing F# proejct system?

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties
+{
+    /// <summary>
+    ///     Provides common well-known Visual Basic project property pages.
+    /// </summary>
+    internal static class FSharpProjectDesignerPage
+    {
+        public static readonly ProjectDesignerPageMetadata Application = new ProjectDesignerPageMetadata(new Guid("{6D2D9B56-2691-4624-A1BF-D07A14594748}"), pageOrder: 0, hasConfigurationCondition: false);
+        public static readonly ProjectDesignerPageMetadata Build = new ProjectDesignerPageMetadata(new Guid("{FAC0A17E-2E70-4211-916A-0D34FB708BFF}"), pageOrder: 1, hasConfigurationCondition: false);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs]: unify these?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/VsContainedLanguageComponentsFactory.cs [hunk 4]
```diff
@@ -14,7 +14,7 @@
 namespace Microsoft.VisualStudio.ProjectSystem.VS.LanguageServices
 {
     [Export(typeof(IVsContainedLanguageComponentsFactory))]
-    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasicOrFSharp)]
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/VsContainedLanguageComponentsFactory.cs]: I reckon we should mark these as "Managed".

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj [hunk 5]
```diff
@@ -33,6 +37,9 @@
     <Compile Update="ProjectSystem\Rules\AdditionalFiles.cs">
       <DependentUpon>AdditionalFiles.xaml</DependentUpon>
     </Compile>
+    <Compile Update="ProjectSystem\Rules\FSharp.cs">
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj]: We get a bug to pull these into appropriate language-specific projects.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/FSharp.BrowseObject.xaml [hunk 6]
```diff
@@ -0,0 +1,99 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!--Copyright, Microsoft Corporation, All rights reserved.-->
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/FSharp.BrowseObject.xaml]: Should we just unify these between C#/VB and F#? Do they have any different properties?

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/FSharp.BrowseObject.xaml]: Actually, why can we just use the C# one? Don't we do that for VB? And just rename the thing to "Compile"

Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/FSharp.BrowseObject.xaml]: Will do as a separate PR since it touches too many files.

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs (hunk 3 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs]: Unify the pages you mean? I think the build page is different for F#.

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Packaging/FSharpProjectSystemPackage.cs (hunk 2 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Packaging/FSharpProjectSystemPackage.cs]: You are right - the old F# PS doesn't seem to register any singlefilegenerators. These likely produce cs files and are incorrect. I've removed them.

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs (hunk 3 above)
Comment (xperiandri) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs]: Because of this?
image

Comment (xperiandri) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs]: Is it XAML? Could you make this page shrinking horizontally?
Textboxes can become wider without any problem so that all the content can be visible.
This with limitation is very inconvenient!

Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs]: Ha - is that the only difference? I just meant that the implementation is different (a separate class with a separate guid in the F# assembly but didn't realize it's that close to C#). Ya it might be nice to consolidate these.

**Comment by enricosada** [comment]
as a note:

- https://github.com/dotnet/netcorecli-fsc/pull/101 to change the guid in f# to use new (i will release this and others needed stuff with `FSharp.NET.Sdk` v`1.0.3`)
- https://github.com/dotnet/netcorecli-fsc/pull/99 will use fsc.exe (so .net full) when msbuild host is .net full (msbuild.exe), like in VS.
- https://github.com/dotnet/netcorecli-fsc/pull/100 update `FSharp.Build` tasks with new propeties, also for .NET full fsc tasks

**Comment by enricosada** [comment]
@srivatsn there are `<Capability` i need to add to `FSharp.NET.Sdk`? if i understand right how that works..

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/FSharpProjectGuidProvider.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Packaging;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    ///     Provides the Visual Basic implementation of <see cref="IItemTypeGuidProvider"/>.
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/FSharpProjectGuidProvider.cs]: `Visual Basic` -> `FSharp`

Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/FSharpProjectGuidProvider.cs]: I guess there are lot of places where Visual Basic or C# will be present instead of F#. As we implement the component we should update these files.

**Comment by srivatsn** [comment]
@enricosada the capabilities are included in the FSharp.DesignTime.targets that I've added in this PR. This targets file needs to be imported by the FSharp SDK. More details in https://github.com/dotnet/roslyn-project-system/issues/1857
